### PR TITLE
feat(flag): Allow Redis client to work in sentinel or cluster mode

### LINKED ISSUE #2772: Question - Does trivy server mode support redis in cluser mode?
I've tried enabling trivy server and connecting it to a redis that was deployed in cluster mode.
The redis-cli works when supplying the `-c` flag which means that cluster is configured properly.

Server was enabled using this command:
`/usr/local/bin/trivy server --cache-backend redis://<redis-hostname>:6379 &`

Scan was triggered using this command:
`/usr/local/bin/trivy image --server http://127.0.0.1:4954 --debug --security-checks vuln <image>`

The error I received seems like an error related to SDKs that do communicate with redis in cluster mode (hence the `MOVED` error response).
So my question is whether trivy supports connections to cluster based redis?
If so, it there some other flag I need to provide?

Error:
```
2022-08-24T13:56:06.036Z	FATAL	image scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.Run
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:367
  - scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.(*runner).scanArtifact
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:227
  - image scan failed:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        /home/runner/work/trivy/trivy/pkg/commands/artifact/run.go:531
  - failed analysis:
    github.com/aquasecurity/trivy/pkg/scanner.Scanner.ScanArtifact
        /home/runner/work/trivy/trivy/pkg/scanner/scan.go:127
  - analyze error:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/image.Artifact.Inspect
        /home/runner/work/trivy/trivy/pkg/fanal/artifact/image/image.go:114
  - failed to store layer: sha256:a021bed5b21226d513871047f19534f8f1bbd87728038dce2e0ca4fc2d044b75 in cache:
    github.com/aquasecurity/trivy/pkg/fanal/artifact/image.Artifact.inspect.func1
        /home/runner/work/trivy/trivy/pkg/fanal/artifact/image/image.go:178
  - unable to store cache on the server:
    github.com/aquasecurity/trivy/pkg/cache.RemoteCache.PutBlob
        /home/runner/work/trivy/trivy/pkg/cache/remote.go:52
  - twirp error internal: unable to store layer info in cache: unable to store blob information in Redis cache (sha256:a021bed5b21226d513871047f19534f8f1bbd87728038dce2e0ca4fc2d044b75): MOVED 425 172.31.25.3:6379
```

### PR DESCRIPTION
**Pull request by bgoareguer**
## Description
Based on work done in #2968, I added support for Redis Sentinel mode:

```bash
# Redis standalone
./trivy image  --cache-backend redis://127.0.0.1:6379 alpine:3.17.0

# Redis Sentinel
./trivy image --cache-backend redis://127.0.0.1:26379 --redis-master-name mymaster alpine:3.17.0

# Redis cluster
./trivy image --cache-backend redis://127.0.0.1:6379,redis://127.0.0.1:6380,redis://127.0.0.1:6381 alpine:3.17.0
```

## Related issues
- Close #2772

## Related PRs
- [ ] #2968

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: docs/docs/vulnerability/examples/cache.md [hunk 1]
```diff
@@ -44,6 +44,29 @@ Two options:
...
+$ trivy server --cache-backend redis://localhost:6379
+
+$ # Redis Sentinel
+$ trivy server --cache-backend redis://localhost:26379 --redis-master-name mymaster
+
+$ # Redis cluster
+$ trivy server -cache-backend redis://redis-0:6379,redis://redis-1:6379,redis://redis-2:6379
+```
+
+If your Redis instance requires authentication, the password can be set in the Redis URL (if multiple Redis URLs are specified, the password will be retrieved only from the last last Redis URL):
```
Comment (AnaisUrlichs) [inline docs/docs/vulnerability/examples/cache.md:64]: You put 'last' twice

File: docs/docs/vulnerability/examples/cache.md [hunk 2]
```diff
@@ -53,4 +76,4 @@ $ trivy server --cache-backend redis://localhost:6379 \
   --redis-key /path/to/key.pem
 ```
 
-TLS option for redis is hidden from Trivy command-line flag, but you still can use it.
\ No newline at end of file
+TLS option for redis is hidden from Trivy command-line flag, but you still can use it.
```
Comment (AnaisUrlichs) [inline docs/docs/vulnerability/examples/cache.md:79]: The TLS option for Redis is hidden from the Trivy command-line flag, but you still can use it.

- make sure to use Redis uppercase everywhere :)

**Review by AnaisUrlichs — COMMENTED** [review]
Added two minor comments to the docs, thank you :)

File: docs/docs/vulnerability/examples/cache.md [hunk 3]
```diff
@@ -44,6 +44,29 @@ Two options:
 $ trivy server --cache-backend redis://localhost:6379
 ```
 
+Trivy supports all Redis deployment options:
+
+- `standalone`
+- `sentinel` (in order to connect to a Redis Sentinel, you need to specify the master name with the `-redis-master-name` option)
```
Comment (AnaisUrlichs) [inline docs/docs/vulnerability/examples/cache.md:50]: just curious, they still call it "master"? Gosh

File: docs/docs/vulnerability/examples/cache.md [hunk 4]
```diff
@@ -44,6 +44,29 @@ Two options:
...
+- `cluster`
+
+```
+$ # Redis standalone
+$ trivy server --cache-backend redis://localhost:6379
+
+$ # Redis Sentinel
+$ trivy server --cache-backend redis://localhost:26379 --redis-master-name mymaster
```
Comment (AnaisUrlichs) [inline docs/docs/vulnerability/examples/cache.md:58]: I cannot get over them calling it "master"

**Review by AnaisUrlichs — APPROVED** [review]
Docs looks good to me

**Comment by bgoareguer** [comment]
@chen-keinan @knqyf263 @itaysk Just wondering if there is anything preventing this PR to be reviewed (except from lack of time)?

**Comment by knqyf263** [comment]
Sorry to be late. We're trying to define the scope of Trivy OSS now. The scalability could be out of scope.

**Comment by bgoareguer** [comment]
For those interested, I ended up following [this](https://yaniv-bhemo.medium.com/how-to-setup-haproxy-for-redis-sentinel-on-kubernetes-37ee70e44464) tutorial.

It works fine with Trivy 0.42.1.

**Comment by claneys** [comment]
Hello,
Sorry comment this closed PR, but I wondering if this PR could be merged ? We are interested with this feature and if we could avoid deploying another component to handle the client part it would be great.
Thanks

**Comment by itaysk** [comment]
@claneys sorry, server mode or Redis is not an area of investment in Trivy open source, as it is a value offering of Aqua commercial offering

**Pull request closed without merging**